_source/_posts/2021-02-01-building-a-github-secrets-scanner.md

```diff
@@ -0,0 +1,219 @@
+---
+layout: blog_post
+title: Building a GitHub Secrets Scanner
+author: vickie-li
+by: contractor
+communities: [python]
+description: "Use Python to build a basic GitHub secrets scanner."
+tags: [python]
+tweets:
+- "Learn the basics of how to scan GitHub for sensitive information."
+- "Would you know if someone checked secrets into GitHub? Learn how to build a tool to check."
+image: blog/building-a-github-secrets-scanner/card.png
+type: awareness
+---
+GitHub reconnaissance is a tactic that attackers use to gather information about their targets. Attackers analyze organizations' GitHub repositories and check for sensitive data that has been accidentally committed or information that could lead to the discovery of a vulnerability.
+For this tutorial, let's build a scanner that automates the GitHub recon process! We will be scanning your GitHub repositories using the method mentioned in the "Tightening Up Your GitHub Security" post.
+We will be working with Python, the GitHub REST API, and the GitPython Library. You will need to obtain a [GitHub personal access token](https://github.com/settings/tokens) and [install GitPython](https://gitpython.readthedocs.io/en/stable/intro.html#installing-gitpython). 
+Make sure that your GitHub personal access token has these permissions: `public_repo`, `read:org`, `read:packages`, `read:user`, `repo:status`
+Before we dive into the code, here are the imports you need:
+```python
+import requests, tempfile, shutil, re
+
+from git import Repo
+from git import NULL_TREE
+```
+## Finding Repositories
+First, we will crawl through your organization's account to find repositories to analyze. We need the GitHub API for this. The GitHub API endpoint is located at `https://api.github.com`.
+You can set the "Accept" header to "application/vnd.github.v3+json" to specify that we are using the REST API v3. And you can provide the access token through an "Authorization" header.
+```python
+GITHUB_ACCESS_TOKEN = "{Insert your token here!}"
+
+headers = {
+    "Authorization": "token {}".format(GITHUB_ACCESS_TOKEN),
+    "Accept": "application/vnd.github.v3+json",
+}
+api_host = "https://api.github.com"
+
+orgs = ["okta"]  # list organizations to analyze
+users = ["vickie"]  # list users to analyze
+```
+List the organization and individual usernames that you want to analyze. Include the usernames of your organization as well as the usernames of your organization's employees. Then, we'll start by looking for additional usernames associated with your organization by querying the GitHub API for organization members.
+```python
+def requests_page(path):
+    resp = requests.Response()
+    try:
+        resp = requests.get(url=path, headers=headers, timeout=15)
+    except:
+        pass
+    return resp.json()
+
+
```

Comment:
Remove extra line break.

Reply:
The extra blank lines are part of standard Python formatting using the `black` idiomatic source code formatting tool.

In this case, the extra line break is part of the PEP-8 standard: https://www.python.org/dev/peps/pep-0008/#blank-lines

> Surround top-level function and class definitions with two blank lines.

Reply:
Interesting. I did not know that. Thanks for the info!